### PR TITLE
feat(sns): Port SNS Governance's periodic tasks from heartbeats to timers

### PR DESCRIPTION
**Pull request by aterga**
This PR ports SNS Governance's periodic tasks from heartbeats to timers, making use of the fact that this canister has recently been ported to IC-CDK. For SNS Governance's periodic tasks, timers are better than heartbeats because they can run less frequently, which should help saving subnet resources.

### CODE REVIEW

File: rs/sns/governance/proto/ic_sns_governance/pb/v1/governance.proto [hunk 1]
```diff
@@ -1471,8 +1471,19 @@ message Governance {
     optional uint64 response_timestamp_seconds = 3;
   }
   CachedUpgradeSteps cached_upgrade_steps = 29;
+
+  // Information about the timers that perform periodic tasks of this Swap canister.
```
Comment (aterga) [inline rs/sns/governance/proto/ic_sns_governance/pb/v1/governance.proto:1475]:
```suggestion
  // Information about the timers that perform periodic tasks of this Governance canister.
```

File: rs/sns/governance/src/gen/ic_sns_governance.pb.v1.rs [hunk 2]
```diff
@@ -1648,6 +1648,9 @@ pub struct Governance {
     pub maturity_modulation: ::core::option::Option<governance::MaturityModulation>,
     #[prost(message, optional, tag = "29")]
     pub cached_upgrade_steps: ::core::option::Option<governance::CachedUpgradeSteps>,
+    /// Information about the timers that perform periodic tasks of this Swap canister.
```
Comment (aterga) [inline rs/sns/governance/src/gen/ic_sns_governance.pb.v1.rs:1651]:
```suggestion
    /// Information about the timers that perform periodic tasks of this Governance canister.
```

File: rs/sns/governance/canister/canister.rs [hunk 3]
```diff
@@ -68,6 +70,15 @@ const STABLE_MEM_BUFFER_SIZE: u32 = 100 * 1024 * 1024; // 100MiB
 
 static mut GOVERNANCE: Option<Governance> = None;
 
+thread_local! {
+    static TIMER_ID: RefCell<Option<TimerId>> = RefCell::new(Default::default());
+}
+
+/// This guarantees that timers cannot be restarted more often than once every 100 intervals.
```
Comment (aterga) [inline rs/sns/governance/canister/canister.rs:77]:
```suggestion
/// This guarantees that timers cannot be restarted more often than once every 300 intervals.
```

File: rs/sns/test_utils/src/itest_helpers.rs [hunk 4]
```diff
@@ -989,6 +989,13 @@ impl SnsCanisters<'_> {
             .await
     }
 
+    pub async fn run_periodic_tasks_now(&self) -> Result<(), String> {
+        self.governance
+            .update_("run_periodic_tasks_now", candid_one, ())
+            .await
+    }
+
+
```
Comment (aterga) [inline rs/sns/test_utils/src/itest_helpers.rs:998]:
```suggestion
```

File: rs/sns/governance/canister/governance_test.did [hunk 5]
```diff
@@ -300,6 +300,12 @@ type Governance = record {
   neurons : vec record { text; Neuron };
   genesis_timestamp_seconds : nat64;
   target_version: opt Version;
+  timers : opt Timers;
```
Comment (max-dfinity) [inline rs/sns/governance/canister/governance_test.did:303]: This should not be in the init arguments... but I guess we haven't split up the API crate from SNS yet...

File: rs/sns/governance/canister/canister.rs [hunk 6]
```diff
@@ -68,6 +70,15 @@ const STABLE_MEM_BUFFER_SIZE: u32 = 100 * 1024 * 1024; // 100MiB
...
 static mut GOVERNANCE: Option<Governance> = None;
 
+thread_local! {
+    static TIMER_ID: RefCell<Option<TimerId>> = RefCell::new(Default::default());
+}
+
+/// This guarantees that timers cannot be restarted more often than once every 300 intervals.
+const RESET_TIMERS_COOL_DOWN_INTERVAL: Duration = Duration::from_secs(600);
+
+const RUN_PERIODIC_TASKS_INTERVAL: Duration = Duration::from_secs(2);
```
Comment (max-dfinity) [inline rs/sns/governance/canister/canister.rs:80]: To get any real benefit from timers, we may want to set it a little higher than this?  Currently we're running heartbeat a bit more than 1x/second I think, and timers are a little bit more expensive.  So I don't know if we're going to see a huge lift from just making it every 2 seconds.  But we can monitor and find out I guess

**Review by max-dfinity — APPROVED** [review]

File: rs/sns/governance/canister/canister.rs (hunk 6 above)
Comment (aterga) [reply rs/sns/governance/canister/canister.rs:80]: Thanks for the feedback. Setting it to 5 seconds for now.

File: rs/sns/governance/canister/governance_test.did (hunk 5 above)
Comment (aterga) [reply rs/sns/governance/canister/governance_test.did:303]: Agreed. So this isn't actionable yet.

File: rs/sns/integration_tests/src/timers.rs [hunk 7]
```diff
@@ -211,6 +222,97 @@ fn test_swap_reset_timers() {
...
+        assert_eq!(
+            last_spawned_timestamp_seconds,
+            last_reset_timestamp_seconds + 1000
+        );
+        last_spawned_timestamp_seconds
+    };
+
+    // Reset the timers.
+    {
+        let payload = Encode!(&swap_pb::ResetTimersRequest {}).unwrap();
```
Comment (max-dfinity) [inline rs/sns/integration_tests/src/timers.rs:275]: I might extract "reset_timers" into a functionm as well as "get_timers",

**Review by max-dfinity — APPROVED** [review]

File: rs/sns/integration_tests/src/timers.rs (hunk 7 above)
Comment (aterga) [reply rs/sns/integration_tests/src/timers.rs:275]: Will do this in a follow-up to avoid invalidating the CI pipeline.

Comment (aterga) [reply rs/sns/integration_tests/src/timers.rs:275]: Done in https://github.com/dfinity/ic/pull/2214

**Pull request merged**